gym/envs/registration.py

```diff
@@ -680,63 +310,190 @@ def make(id: Literal[
     "Ant-v2"
 ], **kwargs) -> Env[np.ndarray, np.ndarray]: ...
 
-# ----------------------------------------
 
 @overload
 def make(id: str, **kwargs) -> Env: ...
+@overload
+def make(id: EnvSpec, **kwargs) -> Env: ...
+
 # fmt: on
-def make(id: str, **kwargs) -> Env:
-    return registry.make(id, **kwargs)
 
 
-def spec(id: str) -> EnvSpec:
-    return registry.spec(id)
+# Global registry of environments. Meant to be accessed through `register` and `make`
+registry: dict[str, EnvSpec] = dict()
+current_namespace: Optional[str] = None
+
+
+def check_spec_register(spec: EnvSpec):
+    global registry, current_namespace
+    if current_namespace is not None:
+        if spec.namespace is not None:
+            logger.warn(
+                f"Custom namespace `{spec.namespace}` is being overridden "
+                f"by namespace `{current_namespace}`. If you are developing a "
+                "plugin you shouldn't specify a namespace in `register` "
+                "calls. The namespace is specified through the "
+                "entry point package metadata."
+            )
+
+    latest_versioned_spec = max(
+        (
+            spec_
+            for spec_ in registry.values()
+            if spec_.name == spec.name and spec_.version is not None
```

Comment:
Do we need to check that the `spec_.namespace == spec.namespace` as there may be two environments, one with a namespace and the other without. i.e. atari where we have `ALE/Pong-v5` and `Pong-v4` but if the versions were the same then could this be an issue?

Reply:
Good catch, fixed